### ISSUE TITLE
TPM/config.yaml resets to "default" after Set-EflowVM

### ISSUE THREAD

**Issue by Thacai** [Azure/iotedge-eflow#72]
**Describe the bug**
After changing example cpuCount, the iotedge service breaks, would appear from logs the previous configuration is completely reset.

**To Reproduce**
1. Stop-EflowVM
2. Set-EflowVM -cpuCount 2 -memoryInMB 1024 (change from 1, could also change memory)
3. Start-EflowVM 

**Expected behavior**
Would expect previous working iotedge, to continue working, after VM cpu count or memory change

**Windows Host OS (please complete the following information):**
 - Edition: Windows 10 Enterprise
 - Version: 20H2
 - Virtual Machine: None

 **Service log**
```
iotedge-user@<snip> [ ~ ]$ sudo systemctl stop iotedge
iotedge-user@<snip> [ ~ ]$ sudo systemctl start iotedge
iotedge-user@<snip> [ ~ ]$ sudo journalctl -u iotedge -r
-- Logs begin at Tue 2021-08-24 09:51:21 UTC, end at Thu 2021-09-02 14:14:30 UTC. --
Sep 02 14:14:26 <snip> systemd[1]: iotedge.service: Failed with result 'exit-code'.
Sep 02 14:14:26 <snip> systemd[1]: iotedge.service: Main process exited, code=exited, s
tatus=153/n/a
Sep 02 14:14:26 <snip> iotedged[2667]: 2021-09-02T14:14:25Z [INFO] (/usr/src/mariner/BUILD/azure-iotedge-1.1.4/
edgelet/hsm-sys/azure-iot-hsm-c/src/hsm_log.c:log_init:41) Initialized logging
Sep 02 14:14:26 <snip> iotedged[2667]: See https://aka.ms/iot-edge-configure-linux for more details.
Sep 02 14:14:26 <snip> iotedged[2667]: Please update config.yaml with the IoT Hub connection information.
Sep 02 14:14:26 <snip> iotedged[2667]: 2021-09-02T14:14:26Z [ERR!] -         caused by:
 Edge device information is required.
Sep 02 14:14:26 <snip> iotedged[2667]: 2021-09-02T14:14:26Z [ERR!] - The daemon could n
ot start up successfully: Could not load settings
Sep 02 14:14:26 <snip> iotedged[2667]: 2021-09-02T14:14:26Z [INFO] - Starting provisioning edge device via manu
al mode using a device connection string...
Sep 02 14:14:26 <snip> iotedged[2667]: 2021-09-02T14:14:26Z [INFO] - Provisioning edge device...
Sep 02 14:14:25 <snip> iotedged[2667]: 2021-09-02T14:14:25Z [INFO] - Finished initializing hsm.
Sep 02 14:14:25 <snip> iotedged[2667]: 2021-09-02T14:14:25Z [INFO] - Initializing hsm...
Sep 02 14:14:25 <snip> iotedged[2667]: 2021-09-02T14:14:25Z [INFO] - Finished configuring provisioning environm
ent variables and certificates.
Sep 02 14:14:25 <snip> iotedged[2667]: 2021-09-02T14:14:25Z [INFO] - Transparent gateway certificates not found
, operating in quick start mode...
Sep 02 14:14:25 <snip> iotedged[2667]: 2021-09-02T14:14:25Z [INFO] - Configuring certificates...
Sep 02 14:14:25 <snip> iotedged[2667]: 2021-09-02T14:14:25Z [INFO] - Configuring /var/lib/iotedge as the home d
irectory.
Sep 02 14:14:25 <snip> iotedged[2667]: 2021-09-02T14:14:25Z [INFO] - Using config file: /etc/iotedge/config.yam
l
Sep 02 14:14:25 <snip> iotedged[2667]: 2021-09-02T14:14:25Z [INFO] - Version - 1.1.4
Sep 02 14:14:25 <snip> iotedged[2667]: 2021-09-02T14:14:25Z [INFO] - Starting Azure IoT Edge Security Daemon
Sep 02 14:14:25 <snip> systemd[1]: Started Azure IoT Edge daemon.
```

**iotedge check**
```
iotedge-user@<snip> [ ~ ]$ sudo iotedge check --verbose
Configuration checks
--------------------
√ config.yaml is well-formed - OK
× config.yaml has well-formed connection string - Error
    Invalid connection string format detected.
    Please check the value of the provisioning.device_connection_string parameter.
        caused by: Edge device information is required.
                   Please update config.yaml with the IoT Hub connection information.
                   See https://aka.ms/iot-edge-configure-linux for more details.
√ container engine is installed and functional - OK
√ config.yaml has correct hostname - OK
× config.yaml has correct URIs for daemon mgmt endpoint - Error
    SocketError - SocketErrorCode (ConnectionRefused) : Connection refused /var/lib/iotedge/mgmt.sock
    One or more errors occurred. (Got bad response: )
        caused by: docker returned exit code: 1, stderr = SocketError - SocketErrorCode (ConnectionRefused) : Connection refused /var/lib/iotedge/mgmt.sock
                   One or more errors occurred. (Got bad response: )
‼ latest security daemon - Warning
    Installed IoT Edge daemon has version 1.1.4 but 1.1.5 is the latest stable version available.
    Please see https://aka.ms/iotedge-update-runtime for update instructions.
√ host time is close to real time - OK
√ container time is close to host time - OK
‼ DNS server - Warning
    Container engine is not configured with DNS server setting, which may impact connectivity to IoT Hub.
    Please see https://aka.ms/iotedge-prod-checklist-dns for best practices.
    You can ignore this warning if you are setting DNS server per module in the Edge deployment.
‼ production readiness: certificates - Warning
    The Edge device is using self-signed automatically-generated development certificates.
    They will expire in 80 days (at 2021-11-22 09:51:15 UTC) causing module-to-module and downstream device communication to fail on an active deployment.
    After the certs have expired, restarting the IoT Edge daemon will trigger it to generate new development certs.
    Please consider using production certificates instead. See https://aka.ms/iotedge-prod-checklist-certs for best practices.
√ production readiness: container engine - OK
√ production readiness: logs policy - OK
‼ production readiness: Edge Agent's storage directory is persisted on the host filesystem - Warning
    The edgeAgent module is not configured to persist its /tmp/edgeAgent directory on the host filesystem.
    Data might be lost if the module is deleted or updated.
    Please see https://aka.ms/iotedge-storage-host for best practices.
‼ production readiness: Edge Hub's storage directory is persisted on the host filesystem - Warning
    The edgeHub module is not configured to persist its /tmp/edgeHub directory on the host filesystem.
    Data might be lost if the module is deleted or updated.
    Please see https://aka.ms/iotedge-storage-host for best practices.

Connectivity checks
-------------------
‼ host can connect to and perform TLS handshake with IoT Hub AMQP port - Warning
    skipping because of previous failures
‼ host can connect to and perform TLS handshake with IoT Hub HTTPS / WebSockets port - Warning
    skipping because of previous failures
‼ host can connect to and perform TLS handshake with IoT Hub MQTT port - Warning
    skipping because of previous failures
‼ container on the default network can connect to IoT Hub AMQP port - Warning
    skipping because of previous failures
‼ container on the default network can connect to IoT Hub HTTPS / WebSockets port - Warning
    skipping because of previous failures
‼ container on the default network can connect to IoT Hub MQTT port - Warning
    skipping because of previous failures
‼ container on the IoT Edge module network can connect to IoT Hub AMQP port - Warning
    skipping because of previous failures
‼ container on the IoT Edge module network can connect to IoT Hub HTTPS / WebSockets port - Warning
    skipping because of previous failures
‼ container on the IoT Edge module network can connect to IoT Hub MQTT port - Warning
    skipping because of previous failures

7 check(s) succeeded.
5 check(s) raised warnings.
2 check(s) raised errors.
9 check(s) were skipped due to errors from other checks.
```

**Comment by ArchanaMSFT** [comment]
Hi @Thacai 
Thank you for reaching out. 
The problem here is not Set-EflowVM. It is about stop and start EflowVm when it is provisioned with TPM.
The default time for TPM to be back is 2 hours. Please check again to see if the modules are running.

The time TPM takes to come back is set as a regkey on the host device. It can be configured.
More information - https://docs.microsoft.com/en-us/previous-versions/windows/it-pro/windows-server-2003/cc782936(v=ws.10). 
Could you please shorten the keepalive setting so that TCP comes back alive sooner?
Please let us know if your modules arent up and running by now. 
Thank you

**Comment by fcabrera23** [comment]
Thanks @ArchanaMSFT  for your response. 

@Thacai, did you try changing the regkey?  I try your repro steps, and couldn't replicate the issue.

Thanks,
Francisco

**Comment by Thacai** [comment]
> Thanks @ArchanaMSFT for your response.
> 
> @Thacai, did you try changing the regkey? I try your repro steps, and couldn't replicate the issue.
> 
> Thanks,
> Francisco

That regkey isnt valid on windows 10, and I doubt it has anything to do with the host, but more likely the TPM2 broker service:

Fully provisioned device(new pc):
![image](https://user-images.githubusercontent.com/4174649/132520586-84833079-c981-485d-a4e7-0ba7ca62ddbc.png)

After the repro steps:
![image](https://user-images.githubusercontent.com/4174649/132520082-48c19fa2-8fd6-4a7e-b2b0-df75ee43ff5d.png)

It's now stuck on retrieving TPM:
![image](https://user-images.githubusercontent.com/4174649/132518434-7e0afa69-bb2e-4f19-88f0-68a23b051ef6.png)

If i look at service logs, it shows it's reprovisioning, but failing due to TPM:
![image](https://user-images.githubusercontent.com/4174649/132519691-20964939-d130-4c7a-9b30-17553d9dcae9.png)


Manual reprovisioning with Provision-EflowVm -provisioningType DpsTPM -scopeId <snip> also fails:
![image](https://user-images.githubusercontent.com/4174649/132519076-f3b5e095-782e-4c2a-a960-df7b930ad049.png)

But the reason I doubt it's the host malfunctioning is due to the following, Disable DpsTPM feature, and re-enable it after VM change(CPU or memory), now the TPM information is being received:
![image](https://user-images.githubusercontent.com/4174649/132520978-fb87659d-bbee-4337-ac5c-b8349f3f39f8.png)

But the iotedge check is still failing:
![image](https://user-images.githubusercontent.com/4174649/132521471-218913a0-2343-46a0-b7bb-767e61a7ecc9.png)

I then run Provision-EflowVm -provisioningType DpsTPM -scopeId <snip> again, note it now doesnt remember it was previously provisioned:
![image](https://user-images.githubusercontent.com/4174649/132521841-8a524ddc-3d25-41e8-a10e-ad7faaf47462.png)

Waiting a bit to get the pending deployment configurations applied, then running iotedge check:
![image](https://user-images.githubusercontent.com/4174649/132522433-356d2b78-d2c0-42bd-a4a9-14bb70f534c9.png)

So my conclusion (based on that I was able to get the TPM to work again, without any change to host) would be that the TPM2 broker service somehow breaks once the VM configuration gets changed, and by manually disabling TPM feature and re-enabling it afterwards makes the TPM feature work "instantly", but at the cost of breaking the config.yaml, requiring a new provisioning. 
And I can say, it seems it wont recover itself even after more than 2h has passed (not fully illustrated in these screenshots).

A change proposal/workaround from my side (if others can reproduce this as well) would be that when a Set-EFlowVM is being run, it should stop the TPM broker service if previously configured, and re-enable afterwards, while persisting the config.yaml, hence it won't require a manual provisioning again (the endorsement key and reg-id doesnt change)

**Comment by fcabrera23** [comment]
Hi @Thacai,

We were able to replicate your issue, and we will provide a fix in our next update (1.1.2109.0). Please check our latest [Release Notes](https://aka.ms/AzEFLOW-Releases).

Thanks,
Francisco
